### DISCUSSION TITLE
How to apply Tailwind styling

### DISCUSSION THREAD

**Discussion by tclasen** [idom-team/idom#757] [Question]
Hello IDOM family. I'm interested to see if there is an example of how to use tailwind to style the application created by IDOM.

**Comment by rmorshea** [comment]
You should be able to link the tailwind stylesheet into your app. Then you'll be able to use all the normal tailwind classes.

```python
from idom import run, component, html

@component
def MyComponent():
    return html._(
        html.link(
            {
                "href": "https://unpkg.com/tailwindcss@^1.0/dist/tailwind.min.css",
                "rel": "stylesheet",
            }
        ),
        html.h1(
            {"className": "text-3xl font-bold underline text-clifford"}, "my heading"
        ),
    )

run(MyComponent)
```

**Reply by Archmonger** [reply]
`src` should be within the properties dictionary (not a kwarg), if I recall.

**Reply by rmorshea** [reply]
Thanks! Corrected.

**Reply by Archmonger** [reply]
comma is required after `html.script({"src": "https://cdn.tailwindcss.com"})`,

And `idom.run` is borked on 0.38.1 [due to a typo in utils.py](https://github.com/idom-team/idom/blob/0177411ddd045e424e5d632a687ca767bd258e6d/src/idom/backend/utils.py#L43). Need to change `idom.server` to `idom.backend`.

```
2022-06-11T16:46:31-0700 | WARNING | You are running a development server. Change this before deploying in production!
Traceback (most recent call last):
  File "c:\Users\Username\Documents\Repositories\idom-testbed\test.py", line 13, in <module>
    run(MyComponent)
  File "C:\Users\Username\Documents\Repositories\idom-testbed\.venv\lib\site-packages\idom\backend\utils.py", line 43, in run
    implementation = implementation or import_module("idom.server.default")
  File "C:\Users\Username\AppData\Local\Programs\Python\Python39\lib\importlib\__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 972, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 984, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'idom.server'
````

**Reply by ethanopp** [reply]
> ```python
>     return html._(
>         html.script({"src": "https://cdn.tailwindcss.com"}),
>         html.h1(
>             {"class": "text-3xl font-bold underline text-clifford"},
>             "my heading"
>         )
>     )
> ```

Is there an updated snipper on how to get this working? I am not getting any errors but it also does not seem to be applying the CSS

**Reply by rmorshea** [reply]
I just updated my answer to reflect the fact that you should use `"className"` instead of `"class"`.

**Reply by ethanopp** [reply]
> I just updated my answer to reflect the fact that you should use `"className"` instead of `"class"`.

That still does not work for me, however this does - and class**Name** is not needed
```Python
        html.link(
            {
                "href": "https://unpkg.com/tailwindcss@^1.0/dist/tailwind.min.css",
                "rel": "stylesheet",
            }
        ),
 ```

**Reply by Archmonger** [reply]
React will handle the "class" keyword if in debug mode, but you'll get an error within your browser console. As such, "className" should always be used.

However that is a separate issue. I'll take a quick peek and see why this example isn't working.

**Reply by Archmonger** [reply]
The example above was using `html.script` instead of `html.link`, and was missing the `rel` property.

```python

from idom import run, component, html


@component
def MyComponent():
    return html._(
        html.link(
            {
                "href": "https://unpkg.com/tailwindcss@^1.0/dist/tailwind.min.css",
                "rel": "stylesheet",
            }
        ),
        html.h1(
            {"className": "text-3xl font-bold underline text-clifford"}, "my heading"
        ),
    )


run(MyComponent)

```

**Reply by ethanopp** [reply]
Okay thanks - will stick with the html.link and use className

**Comment by alonahmias** [comment]
Is there a way to use tailwindcss without a link? To install via npm install and use it? Because i want to deploy it in an offline on prem servers

**Reply by Archmonger** [reply]
A link element is required, however, you can use a local link instead.

Here is an example using [`ReactPy v2.0.0b2`](https://reactive-python.github.io/reactpy/learn/creating-a-react-app/) and [`ServeStatic`](https://archmonger.github.io/ServeStatic/latest/asgi/)


```bash
> pip install reactpy[asgi]==2.0.0b2 servestatic uvicorn[standard] -U
```

```python
# FILENAME: main.py
from reactpy import component, html
from reactpy.executors.asgi import ReactPy
from servestatic import ServeStaticASGI

@component
def hello_world():
    return html.button({"className": "bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded"}, "Hello, World!")

app = ReactPy(
    hello_world,
    html_head=html.head(
        html.title("My App"),
        html.meta({"charset": "UTF-8"}),
        html.link({"href": "/static/tailwind.min.css", "rel": "stylesheet"}),
    ),
)

# Note: This line assumes you downloaded a tailwind distributable from NPM and placed it in a directory named "static"
app = ServeStaticASGI(my_app, root="static", prefix="/static/")
```

```bash
> uvicorn main:app
```
